### PR TITLE
Pattern matching compilation

### LINKED ISSUE #1225: Add coverage checking
# Description

Coverage checking must be performed after type checking. The algorithm expects all the patterns in clauses to be valid, meaning that every pattern covers at least one value of the corresponding argument type of the function.

# Checklist

- [ ] Add a flag to disable the coverage check
- [ ] Add a keyword to disable exhaustiveness, possibly called `total`
- [ ] Add tests:
  - [ ] Negative tests
  - [ ] Positive tests
  - [ ] Shell tests 

- Safeguards:
  - [ ] To carry on the analysis, it comes in handy to compute for each pattern row their set of "compatible" rows. Two rows are incompatible if they have no instance in common.
  - [ ] Delete any row that is "subsumed" by another row.
  - [ ] We announce exhaustiveness as soon as we find a row full of wildcards.

Checks:

- Exhaustiveness: (meaning that the pattern _ does not contribute to match more values):
  
  - [ ] Error: The pattern-matching is not exhaustive.
    - [ ] Show non-matching values as part of the error message
- [ ] Exact split tree

### LINKED ISSUE #1798: Efficient compilation of pattern matching
We should implement one of the standard algorithms in the literature, probably the one from SPJ's book. This should also do coverage checking (#1225). The first version of pattern matching compilation implemented for #1531 will be very inefficient and won't do coverage checking.

Literature
-------------
* Philip Wadler, Efficient compilation of pattern matching, in: Simon Peyton Jones, [The implementation of functional programming languages](https://www.microsoft.com/en-us/research/uploads/prod/1987/01/slpj-book-1987.pdf), Chapter 5.
* Luc Maranget, [Compiling pattern matching to good decision trees](http://moscova.inria.fr/~maranget/papers/ml05e-maranget.pdf)
* Jules Jacobs, [How to compile pattern matching](https://julesjacobs.com/notes/patternmatching/patternmatching.pdf)
* Luc Maranget, [Warnings for pattern matching](http://moscova.inria.fr/~maranget/papers/warn/warn.pdf)

### LINKED ISSUE #1926: out of memory errors when running WASM compiled bench examples
The runtime crash can be reproduced with the merge sort and fibonacci bench tests:

https://github.com/anoma/juvix/blob/8d7e669f7499b4122d9b1a9baf42828d1c163c7a/tests/benchmark/mergesort/juvix/mergesort.juvix
https://github.com/anoma/juvix/blob/8d7e669f7499b4122d9b1a9baf42828d1c163c7a/tests/benchmark/fibonacci/juvix/fibonacci.juvix

```
$ cd tests/benchmark/mergesort/juvix
$ juvix compile -t wasm32-wasi mergesort.juvix
$ wasmer mergesort.wasm
error: out of memory
error: failed to run `mergesort.wasm`
│   1: RuntimeError: unreachable
╰─▶ 2: unreachable
```

When running with wasmtime instead of wasmer there's a bit more information:

```
$ wasmtime mergesort.wasm
error: out of memory
Error: failed to run main module `mergesort.wasm`

Caused by:
    0: failed to invoke command default
    1: error while executing at wasm backtrace:
           0: 0x2795 - <unknown>!<wasm function 14>
           1: 0x20c3 - <unknown>!<wasm function 11>
           2: 0x20af - <unknown>!<wasm function 10>
           3: 0x1963 - <unknown>!<wasm function 3>
    2: wasm trap: wasm `unreachable` instruction executed
```

There are no errors when running the same example after compiling with the native target.

### PR DESCRIPTION
**Pull request by lukaszcz**
This implements a basic version of the algorithm from: Luc Maranget, [Compiling pattern matching to good decision trees](http://moscova.inria.fr/~maranget/papers/ml05e-maranget.pdf). No heuristics are used - the first column is always chosen.

* Closes #1798 
* Closes #1225 
* Closes #1926 
* Adds a global `--no-coverage` option which turns off coverage checking in favour of generating runtime failures
* Changes the representation of Match patterns in JuvixCore to achieve a more streamlined implementation
* Adds options to the Core pipeline

### CODE REVIEW

File: src/Juvix/Compiler/Core/Transformation/MatchToCase.hs [hunk 1]
```diff
@@ -1,338 +1,314 @@
...
+    compileBranch err bindersNum vs col matrix tag = do
+      tab <- ask
+      let ci = fromJust $ HashMap.lookup tag (tab ^. infoConstructors)
+          paramsNum = getTypeParamsNum tab (ci ^. constructorType)
+          argsNum = length (typeArgs (ci ^. constructorType))
+          bindersNum' = bindersNum + argsNum
+          vs' = [bindersNum .. bindersNum + argsNum - 1]
+          err' args =
+            err
+              (parensIf (argsNum > paramsNum) (foldl' (<+>) (pretty (ci ^. constructorName)) (drop paramsNum (take argsNum args))) : drop argsNum args)
```
Comment (janmasrovira) [inline src/Juvix/Compiler/Core/Transformation/MatchToCase.hs:232]: idem

File: src/Juvix/Compiler/Core/Transformation/MatchToCase.hs [hunk 2]
```diff
@@ -1,338 +1,314 @@
...
+              (rbcs, acc) <- a
+              let bcs = map (\b -> BCRemove (BinderRemove b (error "pattern compiler: dependently typed pattern"))) (getPatternExtraBinders pat)
+                  bc = BCRemove (BinderRemove (getPatternBinder pat) (mkVar' 0))
+              binder <- overM binderType (goMatchToCase (recur . revAppend rbcs)) (getPatternBinder pat)
+              return (revAppend bcs (bc : BCAdd 1 : rbcs), binder : acc)
+          )
+          (return (matrix !! argPatsIndex ^. patternRowBinderChangesRev, []))
+          argPats
+      where
+        argPats =
```
Comment (janmasrovira) [inline src/Juvix/Compiler/Core/Transformation/MatchToCase.hs:259]: reduce code duplication. I suggest
```
....
          (return (matrix !! (argPatsIx ^. indexedIx) ^. patternRowBinderChangesRev, []))
          (argPatsIx ^. indexedThing)
      where
        argPatsIx :: Indexed [Pattern]
        argPatsIx = fromJust (firstJust (mapM getArgs) (indexFrom 0 col))
          where
          getArgs :: Pattern -> Maybe [Pattern]
          getArgs = \case
              PatConstr PatternConstr {..}
                | _patternConstrTag == tag -> Just _patternConstrArgs
              _ -> Nothing
```

File: src/Juvix/Compiler/Core/Transformation/MatchToCase.hs [hunk 3]
```diff
@@ -1,338 +1,314 @@
...
+              col
+
+    getBranchMatrix :: [Pattern] -> PatternMatrix -> Tag -> Sem r PatternMatrix
+    getBranchMatrix col matrix tag = do
+      tab <- ask
+      let ci = fromJust $ HashMap.lookup tag (tab ^. infoConstructors)
+          argtys = typeArgs (ci ^. constructorType)
+          argsNum = length argtys
+      return $
+        catMaybes $
```
Comment (janmasrovira) [inline src/Juvix/Compiler/Core/Transformation/MatchToCase.hs:289]: hard to read expression. Suggestion:
```
      let ci = fromJust $ HashMap.lookup tag (tab ^. infoConstructors)
          argtys = typeArgs (ci ^. constructorType)
          argsNum = length argtys
          helper :: PatternRow -> Pattern -> Maybe PatternRow
          helper row = \case
            PatConstr PatternConstr {..}
              | _patternConstrTag == tag ->
                  Just $
                    row
                      { _patternRowPatterns =
                          _patternConstrArgs ++ row ^. patternRowPatterns,
                        _patternRowBinderChangesRev = BCAdd argsNum : row ^. patternRowBinderChangesRev
                      }
            PatWildcard {} ->
              Just $
                row
                  { _patternRowPatterns =
                      map (PatWildcard . PatternWildcard mempty . Binder "_" Nothing) argtys
                        ++ row ^. patternRowPatterns,
                    _patternRowBinderChangesRev = BCAdd argsNum : row ^. patternRowBinderChangesRev,
                    _patternRowIgnoredPatternsNum = argsNum + row ^. patternRowIgnoredPatternsNum
                  }
            _ -> Nothing
      return (catMaybes [helper row pat | (pat, row) <- zipExact col matrix])

```

File: src/Juvix/Compiler/Core/Transformation/MatchToCase.hs [hunk 4]
```diff
@@ -1,338 +1,314 @@
...
+                  CoreError
+                    { _coreErrorMsg = "Pattern matching not exhaustive. Example pattern sequence not matched: " <> show (ppOutput $ err (repeat "_")),
+                      _coreErrorNode = Nothing,
+                      _coreErrorLoc = fromMaybe defaultLoc (getNodeLocation node)
+                    }
+            | otherwise ->
+                return $
+                  mkBuiltinApp' OpFail [mkConstant' (ConstString ("Pattern sequence not matched: " <> pat))]
+        where
+          pat = show (ppOutput $ err (repeat "_"))
```
Comment (janmasrovira) [inline src/Juvix/Compiler/Core/Transformation/MatchToCase.hs:97]: could we have `replicate ? "_"` instead of `repeat`?

**Review by janmasrovira — COMMENTED** [review]
A few minor style comments

File: app/Commands/Dev/Core/Compile/Base.hs [hunk 5]
```diff
@@ -4,8 +4,7 @@ import Commands.Base
 import Commands.Dev.Core.Compile.Options
 import Commands.Extra.Compile qualified as Compile
 import Data.Text.IO qualified as TIO
-import Juvix.Compiler.Asm.Options qualified as Asm
-import Juvix.Compiler.Asm.Pretty qualified as Pretty
+import Juvix.Compiler.Asm.Pretty qualified as Core
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply app/Commands/Dev/Core/Compile/Base.hs:7]: Should be. Done.

File: app/Commands/Extra/Paths.hs [hunk 6]
```diff
@@ -5,3 +5,6 @@ import Juvix.Prelude
 -- | imaginary file path for error messages in the repl.
 replPath :: Path Abs File
 replPath = $(mkAbsFile "/<repl>")
+
+gebReplPath :: Path Abs File
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply app/Commands/Extra/Paths.hs:9]: I don't think it is. I just copy-pasted it from the Geb REPL to have it in the same file as the analogous variable for Juvix REPL.

File: src/Juvix/Compiler/Backend.hs [hunk 7]
```diff
@@ -1,9 +1,10 @@
 module Juvix.Compiler.Backend where
 
+import GHC.Base (maxInt)
 import Juvix.Prelude
 
-data Target = TargetCWasm32Wasi | TargetCNative64
-  deriving stock (Data)
+data Target = TargetCWasm32Wasi | TargetCNative64 | TargetGeb | TargetCore | TargetAsm
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Backend.hs:6]: Done

File: src/Juvix/Compiler/Core/Transformation/MatchToCase.hs [hunk 8]
```diff
@@ -1,338 +1,314 @@
...
+          mockFile = $(mkAbsFile "/match-to-case")
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Transformation/MatchToCase.hs:98]: You mean improving the printing of patterns (add colors) or the location? The location of the error is already correct as long as the match has a location set, and it does if it originates from a function in a Juvix file.

File: src/Juvix/Compiler/Core/Transformation/MatchToCase.hs [hunk 9]
```diff
@@ -1,338 +1,314 @@
...
+    compile :: ([Doc Ann] -> Doc Ann) -> Level -> [Level] -> PatternMatrix -> Sem r Node
+    compile err bindersNum vs matrix = case matrix of
+      [] -> do
+        -- The matrix has no rows -- matching fails (Section 4, case 1).
+        fCoverage <- asks (^. optCheckCoverage)
+        if
+            | fCoverage ->
+                throw
+                  CoreError
+                    { _coreErrorMsg = "Pattern matching not exhaustive. Example pattern sequence not matched: " <> show (ppOutput $ err (repeat "_")),
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Transformation/MatchToCase.hs:89]: Done

File: src/Juvix/Compiler/Core/Transformation/MatchToCase.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (janmasrovira) [reply src/Juvix/Compiler/Core/Transformation/MatchToCase.hs:98]: so is it impossible that the user sees `"/match-to-case"` if it compiles from Juvix source?
Either way, we should clarify it in a comment

File: src/Juvix/Compiler/Core/Transformation/MatchToCase.hs [hunk 10]
```diff
@@ -1,338 +1,314 @@
...
+                  case pat of
+                    PatWildcard {} -> Just row
+                    _ -> Nothing
+              )
+              col
+              matrix
+        err' tab args =
+          case mtag of
+            Just tag ->
+              err (parensIf (argsNum > 0) (foldl' (<+>) (pretty (ci ^. constructorName)) (replicate argsNum "_")) : args)
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Transformation/MatchToCase.hs:211]: No, `foldl` works fine here. You can see on `tests/Compilation/negative/test001.juvix`.

I agree using `hsep` is better, but the way you propose would add an extra space when there are no arguments (actually, I has that before).

I changed it to: `hsep (pretty (ci ^. constructorName) : replicate argsNum "_")`

File: src/Juvix/Compiler/Core/Transformation/MatchToCase.hs (hunk 1 above)
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Transformation/MatchToCase.hs:232]: Done

**Comment by lukaszcz** [comment]
It is possible, but it's because location information is not preserved when
you translate a lambda or a case expression. This requires changes to
Internal and Internal-to-Core. Right now, the location information in
Internal is associated with names, so anonymous lambdas and case
expressions don't have it (or I didn't know how to fetch it).

You have `getNodeLocation node` above, so I think it's clear that if a node
has a location info set, then this will not appear.

File: src/Juvix/Compiler/Core/Transformation/MatchToCase.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Transformation/MatchToCase.hs:98]: See also issue #1846 which is about preserving binder names and locations. In addition, we need to preserve location info for all matches that are generated by Internal-to-Core.

[the remark this answers is not in the document]
Comment (janmasrovira) [reply src/Juvix/Compiler/Core/Transformation/MatchToCase.hs:98]: Then #1846 should have higher priority I'd say. Until now no user-facing errors could occur after Internal-to-Core but this pr changes that.
Not having location in an error message should be considred a bug

File: src/Juvix/Compiler/Core/Transformation/MatchToCase.hs (hunk 4 above)
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Transformation/MatchToCase.hs:97]: Done

**Review by janmasrovira — APPROVED** [review]

**Pull request merged**